plugin/pkg/auth: add OpenID Connect token authenticator

yifan-gu opened the pull request:
This PR implements the token authenticator interface for the OIDC protocols. It also adds two flags ("--oidc-endpoint", "--oidc-client-id") for apiserver
to enable OIDC authentication.

It mainly uses [go-oidc packages](https://github.com/coreos/go-oidc). It uses the [oidc client](http://godoc.org/github.com/coreos/go-oidc/oidc#Client) to sync and cache the pub keys automatically. Also it will verify the token using the pub keys.

/cc @erictune @liggitt @smarterclayton

vmarmol commented:
Thank you very much for the pull request - we sincerely appreciate the work that goes into sending us a change.

We are currently in a feature-freeze and general code slush as we work toward our 1.0 release.  Our initial triage indicates that this is a non-critical change at this time, so we'd like to leave it open and revisit it after 1.0.  We expect the freeze to wrap up in early to mid July.

If you feel that this triage is incorrect, please respond and let us know why.  We're currently accepting changes that meet any of the following criteria:
   1. Changes directly related to v1.0 issues: https://github.com/GoogleCloudPlatform/kubernetes/issues?q=is%3Aopen+is%3Aissue+milestone%3Av1.0
   2. Changes addressing P0 bugs: https://github.com/GoogleCloudPlatform/kubernetes/issues?q=is%3Aopen+is%3Aissue+label%3Apriority%2FP0
   3. Changes to supported platforms to bring them up to par with the primary development platforms
   4. Docs, user experience, and test related fixes

Thanks again for your contribution!  The community is really what has made this project so exciting.

yifan-gu commented:
Think this is a post 1.0 feature.

smarterclayton commented:
@liggitt

On Jul 8, 2015, at 7:05 PM, Yifan Gu <notifications@github.com> wrote:

This PR implements the token authenticator interface for the OpenID
protocols. It also adds two flags ("--openid-endpoint",
"--openid-client-id") for apiserver
to enable openid authentication.

It mainly uses go-oidc packages <https://github.com/coreos/go-oidc>. It
uses the oidc client
<http://godoc.org/github.com/coreos/go-oidc/oidc#Client> to sync and cache
the pub keys automatically. Also it will verify the token using the pub
keys.

/cc @erictune <https://github.com/erictune> @liggitt
<https://github.com/liggitt> @smarterclayton
<https://github.com/smarterclayton>
------------------------------
You can view, comment on, or merge this pull request online at:

  https://github.com/GoogleCloudPlatform/kubernetes/pull/10957
Commit Summary

   - plugin/pkg/auth: add openid token authenticator.
   - Godeps: add Godpes for go-oidc.
   - plugin/pkg/auth: add tests for openid authenticator.

File Changes

   - *M* Godeps/Godeps.json
   <https://github.com/GoogleCloudPlatform/kubernetes/pull/10957/files#diff-0>
   (50)
   - *A* Godeps/_workspace/src/github.com/coreos/go-oidc/http/client.go
   <https://github.com/GoogleCloudPlatform/kubernetes/pull/10957/files#diff-1>
   (51)
   - *A* Godeps/_workspace/src/github.com/coreos/go-oidc/http/http.go
   <https://github.com/GoogleCloudPlatform/kubernetes/pull/10957/files#diff-2>
   (159)
   - *A* Godeps/_workspace/src/github.com/coreos/go-oidc/http/http_test.go
   <https://github.com/GoogleCloudPlatform/kubernetes/pull/10957/files#diff-3>
   (380)
   - *A* Godeps/_workspace/src/github.com/coreos/go-oidc/http/middleware.go
   <https://github.com/GoogleCloudPlatform/kubernetes/pull/10957/files#diff-4>
   (14)
   - *A* Godeps/_workspace/src/github.com/coreos/go-oidc/jose/claims.go
   <https://github.com/GoogleCloudPlatform/kubernetes/pull/10957/files#diff-5>
   (79)
   - *A* Godeps/_workspace/src/github.com/coreos/go-oidc/jose/claims_test.go
   <https://github.com/GoogleCloudPlatform/kubernetes/pull/10957/files#diff-6>
   (240)
   - *A* Godeps/_workspace/src/github.com/coreos/go-oidc/jose/jose.go
   <https://github.com/GoogleCloudPlatform/kubernetes/pull/10957/files#diff-7>
   (61)
   - *A* Godeps/_workspace/src/github.com/coreos/go-oidc/jose/jwk.go
   <https://github.com/GoogleCloudPlatform/kubernetes/pull/10957/files#diff-8>
   (131)
   - *A* Godeps/_workspace/src/github.com/coreos/go-oidc/jose/jwk_test.go
   <https://github.com/GoogleCloudPlatform/kubernetes/pull/10957/files#diff-9>
   (64)
   - *A* Godeps/_workspace/src/github.com/coreos/go-oidc/jose/jws.go
   <https://github.com/GoogleCloudPlatform/kubernetes/pull/10957/files#diff-10>
   (51)
   - *A* Godeps/_workspace/src/github.com/coreos/go-oidc/jose/jws_test.go
   <https://github.com/GoogleCloudPlatform/kubernetes/pull/10957/files#diff-11>
   (74)
   - *A* Godeps/_workspace/src/github.com/coreos/go-oidc/jose/jwt.go
   <https://github.com/GoogleCloudPlatform/kubernetes/pull/10957/files#diff-12>
   (84)
   - *A* Godeps/_workspace/src/github.com/coreos/go-oidc/jose/jwt_test.go
   <https://github.com/GoogleCloudPlatform/kubernetes/pull/10957/files#diff-13>
   (94)
   - *A* Godeps/_workspace/src/github.com/coreos/go-oidc/jose/sig.go
   <https://github.com/GoogleCloudPlatform/kubernetes/pull/10957/files#diff-14>
   (25)
   - *A* Godeps/_workspace/src/github.com/coreos/go-oidc/jose/sig_hmac.go
   <https://github.com/GoogleCloudPlatform/kubernetes/pull/10957/files#diff-15>
   (68)
   - *A*
   Godeps/_workspace/src/github.com/coreos/go-oidc/jose/sig_hmac_test.go
   <https://github.com/GoogleCloudPlatform/kubernetes/pull/10957/files#diff-16>
   (85)
   - *A* Godeps/_workspace/src/github.com/coreos/go-oidc/jose/sig_rsa.go
   <https://github.com/GoogleCloudPlatform/kubernetes/pull/10957/files#diff-17>
   (68)
   - *A* Godeps/_workspace/src/github.com/coreos/go-oidc/key/key.go
   <https://github.com/GoogleCloudPlatform/kubernetes/pull/10957/files#diff-18>
   (139)
   - *A* Godeps/_workspace/src/github.com/coreos/go-oidc/key/key_test.go
   <https://github.com/GoogleCloudPlatform/kubernetes/pull/10957/files#diff-19>
   (68)
   - *A* Godeps/_workspace/src/github.com/coreos/go-oidc/key/manager.go
   <https://github.com/GoogleCloudPlatform/kubernetes/pull/10957/files#diff-20>
   (99)
   - *A* Godeps/_workspace/src/github.com/coreos/go-oidc/key/manager_test.go
   <https://github.com/GoogleCloudPlatform/kubernetes/pull/10957/files#diff-21>
   (225)
   - *A* Godeps/_workspace/src/github.com/coreos/go-oidc/key/repo.go
   <https://github.com/GoogleCloudPlatform/kubernetes/pull/10957/files#diff-22>
   (45)
   - *A* Godeps/_workspace/src/github.com/coreos/go-oidc/key/rotate.go
   <https://github.com/GoogleCloudPlatform/kubernetes/pull/10957/files#diff-23>
   (131)
   - *A* Godeps/_workspace/src/github.com/coreos/go-oidc/key/rotate_test.go
   <https://github.com/GoogleCloudPlatform/kubernetes/pull/10957/files#diff-24>
   (240)
   - *A* Godeps/_workspace/src/github.com/coreos/go-oidc/key/sync.go
   <https://github.com/GoogleCloudPlatform/kubernetes/pull/10957/files#diff-25>
   (91)
   - *A* Godeps/_workspace/src/github.com/coreos/go-oidc/key/sync_test.go
   <https://github.com/GoogleCloudPlatform/kubernetes/pull/10957/files#diff-26>
   (205)
   - *A* Godeps/_workspace/src/github.com/coreos/go-oidc/oauth2/error.go
   <https://github.com/GoogleCloudPlatform/kubernetes/pull/10957/files#diff-27>
   (38)
   - *A*
   Godeps/_workspace/src/github.com/coreos/go-oidc/oauth2/error_test.go
   <https://github.com/GoogleCloudPlatform/kubernetes/pull/10957/files#diff-28>
   (79)
   - *A* Godeps/_workspace/src/github.com/coreos/go-oidc/oauth2/oauth2.go
   <https://github.com/GoogleCloudPlatform/kubernetes/pull/10957/files#diff-29>
   (306)
   - *A*
   Godeps/_workspace/src/github.com/coreos/go-oidc/oauth2/oauth2_test.go
   <https://github.com/GoogleCloudPlatform/kubernetes/pull/10957/files#diff-30>
   (258)
   - *A* Godeps/_workspace/src/github.com/coreos/go-oidc/oidc/README.md
   <https://github.com/GoogleCloudPlatform/kubernetes/pull/10957/files#diff-31>
   (1)
   - *A* Godeps/_workspace/src/github.com/coreos/go-oidc/oidc/client.go
   <https://github.com/GoogleCloudPlatform/kubernetes/pull/10957/files#diff-32>
   (296)
   - *A* Godeps/_workspace/src/github.com/coreos/go-oidc/oidc/client_test.go
   <https://github.com/GoogleCloudPlatform/kubernetes/pull/10957/files#diff-33>
   (367)
   - *A* Godeps/_workspace/src/github.com/coreos/go-oidc/oidc/identity.go
   <https://github.com/GoogleCloudPlatform/kubernetes/pull/10957/files#diff-34>
   (44)
   - *A*
   Godeps/_workspace/src/github.com/coreos/go-oidc/oidc/identity_test.go
   <https://github.com/GoogleCloudPlatform/kubernetes/pull/10957/files#diff-35>
   (113)
   - *A* Godeps/_workspace/src/github.com/coreos/go-oidc/oidc/interface.go
   <https://github.com/GoogleCloudPlatform/kubernetes/pull/10957/files#diff-36>
   (3)
   - *A* Godeps/_workspace/src/github.com/coreos/go-oidc/oidc/key.go
   <https://github.com/GoogleCloudPlatform/kubernetes/pull/10957/files#diff-37>
   (57)
   - *A* Godeps/_workspace/src/github.com/coreos/go-oidc/oidc/provider.go
   <https://github.com/GoogleCloudPlatform/kubernetes/pull/10957/files#diff-38>
   (260)
   - *A*
   Godeps/_workspace/src/github.com/coreos/go-oidc/oidc/provider_test.go
   <https://github.com/GoogleCloudPlatform/kubernetes/pull/10957/files#diff-39>
   (466)
   - *A* Godeps/_workspace/src/github.com/coreos/go-oidc/oidc/transport.go
   <https://github.com/GoogleCloudPlatform/kubernetes/pull/10957/files#diff-40>
   (79)
   - *A*
   Godeps/_workspace/src/github.com/coreos/go-oidc/oidc/transport_test.go
   <https://github.com/GoogleCloudPlatform/kubernetes/pull/10957/files#diff-41>
   (167)
   - *A* Godeps/_workspace/src/github.com/coreos/go-oidc/oidc/util.go
   <https://github.com/GoogleCloudPlatform/kubernetes/pull/10957/files#diff-42>
   (59)
   - *A* Godeps/_workspace/src/github.com/coreos/go-oidc/oidc/util_test.go
   <https://github.com/GoogleCloudPlatform/kubernetes/pull/10957/files#diff-43>
   (57)
   - *A*
   Godeps/_workspace/src/github.com/coreos/go-oidc/oidc/verification.go
   <https://github.com/GoogleCloudPlatform/kubernetes/pull/10957/files#diff-44>
   (168)
   - *A*
   Godeps/_workspace/src/github.com/coreos/go-oidc/oidc/verification_test.go
   <https://github.com/GoogleCloudPlatform/kubernetes/pull/10957/files#diff-45>
   (297)
   - *A* Godeps/_workspace/src/github.com/coreos/go-oidc/pkg/crypto/aes.go
   <https://github.com/GoogleCloudPlatform/kubernetes/pull/10957/files#diff-46>
   (94)
   - *A*
   Godeps/_workspace/src/github.com/coreos/go-oidc/pkg/crypto/aes_test.go
   <https://github.com/GoogleCloudPlatform/kubernetes/pull/10957/files#diff-47>
   (93)
   - *A* Godeps/_workspace/src/github.com/coreos/go-oidc/pkg/crypto/rand.go
   <https://github.com/GoogleCloudPlatform/kubernetes/pull/10957/files#diff-48>
   (17)
   - *A* Godeps/_workspace/src/github.com/coreos/go-oidc/pkg/net/url.go
   <https://github.com/GoogleCloudPlatform/kubernetes/pull/10957/files#diff-49>
   (20)
   - *A* Godeps/_workspace/src/github.com/coreos/go-oidc/pkg/net/url_test.go
   <https://github.com/GoogleCloudPlatform/kubernetes/pull/10957/files#diff-50>
   (30)
   - *A* Godeps/_workspace/src/github.com/coreos/go-oidc/pkg/time/backoff.go
   <https://github.com/GoogleCloudPlatform/kubernetes/pull/10957/files#diff-51>
   (15)
   - *A*
   Godeps/_workspace/src/github.com/coreos/go-oidc/pkg/time/backoff_test.go
   <https://github.com/GoogleCloudPlatform/kubernetes/pull/10957/files#diff-52>
   (52)
   - *A* Godeps/_workspace/src/github.com/coreos/pkg/capnslog/README.md
   <https://github.com/GoogleCloudPlatform/kubernetes/pull/10957/files#diff-53>
   (38)
   - *A*
   Godeps/_workspace/src/github.com/coreos/pkg/capnslog/example/hello_dolly.go
   <https://github.com/GoogleCloudPlatform/kubernetes/pull/10957/files#diff-54>
   (59)
   - *A* Godeps/_workspace/src/github.com/coreos/pkg/capnslog/formatters.go
   <https://github.com/GoogleCloudPlatform/kubernetes/pull/10957/files#diff-55>
   (63)
   - *A*
   Godeps/_workspace/src/github.com/coreos/pkg/capnslog/glog_formatter.go
   <https://github.com/GoogleCloudPlatform/kubernetes/pull/10957/files#diff-56>
   (95)
   - *A* Godeps/_workspace/src/github.com/coreos/pkg/capnslog/log_hijack.go
   <https://github.com/GoogleCloudPlatform/kubernetes/pull/10957/files#diff-57>
   (39)
   - *A* Godeps/_workspace/src/github.com/coreos/pkg/capnslog/logmap.go
   <https://github.com/GoogleCloudPlatform/kubernetes/pull/10957/files#diff-58>
   (240)
   - *A* Godeps/_workspace/src/github.com/coreos/pkg/capnslog/pkg_logger.go
   <https://github.com/GoogleCloudPlatform/kubernetes/pull/10957/files#diff-59>
   (158)
   - *A*
   Godeps/_workspace/src/github.com/coreos/pkg/capnslog/syslog_formatter.go
   <https://github.com/GoogleCloudPlatform/kubernetes/pull/10957/files#diff-60>
   (65)
   - *A* Godeps/_workspace/src/github.com/coreos/pkg/health/README.md
   <https://github.com/GoogleCloudPlatform/kubernetes/pull/10957/files#diff-61>
   (11)
   - *A* Godeps/_workspace/src/github.com/coreos/pkg/health/health.go
   <https://github.com/GoogleCloudPlatform/kubernetes/pull/10957/files#diff-62>
   (111)
   - *A* Godeps/_workspace/src/github.com/coreos/pkg/health/health_test.go
   <https://github.com/GoogleCloudPlatform/kubernetes/pull/10957/files#diff-63>
   (198)
   - *A* Godeps/_workspace/src/github.com/coreos/pkg/httputil/README.md
   <https://github.com/GoogleCloudPlatform/kubernetes/pull/10957/files#diff-64>
   (13)
   - *A* Godeps/_workspace/src/github.com/coreos/pkg/httputil/json.go
   <https://github.com/GoogleCloudPlatform/kubernetes/pull/10957/files#diff-65>
   (27)
   - *A* Godeps/_workspace/src/github.com/coreos/pkg/httputil/json_test.go
   <https://github.com/GoogleCloudPlatform/kubernetes/pull/10957/files#diff-66>
   (56)
   - *A* Godeps/_workspace/src/github.com/jonboulle/clockwork/.gitignore
   <https://github.com/GoogleCloudPlatform/kubernetes/pull/10957/files#diff-67>
   (25)
   - *A* Godeps/_workspace/src/github.com/jonboulle/clockwork/.travis.yml
   <https://github.com/GoogleCloudPlatform/kubernetes/pull/10957/files#diff-68>
   (3)
   - *A* Godeps/_workspace/src/github.com/jonboulle/clockwork/LICENSE
   <https://github.com/GoogleCloudPlatform/kubernetes/pull/10957/files#diff-69>
   (201)
   - *A* Godeps/_workspace/src/github.com/jonboulle/clockwork/README.md
   <https://github.com/GoogleCloudPlatform/kubernetes/pull/10957/files#diff-70>
   (61)
   - *A* Godeps/_workspace/src/github.com/jonboulle/clockwork/clockwork.go
   <https://github.com/GoogleCloudPlatform/kubernetes/pull/10957/files#diff-71>
   (164)
   - *A*
   Godeps/_workspace/src/github.com/jonboulle/clockwork/clockwork_test.go
   <https://github.com/GoogleCloudPlatform/kubernetes/pull/10957/files#diff-72>
   (120)
   - *A*
   Godeps/_workspace/src/github.com/jonboulle/clockwork/example_test.go
   <https://github.com/GoogleCloudPlatform/kubernetes/pull/10957/files#diff-73>
   (49)
   - *M* cmd/kube-apiserver/app/server.go
   <https://github.com/GoogleCloudPlatform/kubernetes/pull/10957/files#diff-74>
   (6)
   - *M* pkg/apiserver/authn.go
   <https://github.com/GoogleCloudPlatform/kubernetes/pull/10957/files#diff-75>
   (21)
   - *A* plugin/pkg/auth/authenticator/token/openid/openid.go
   <https://github.com/GoogleCloudPlatform/kubernetes/pull/10957/files#diff-76>
   (112)
   - *A* plugin/pkg/auth/authenticator/token/openid/openid_test.go
   <https://github.com/GoogleCloudPlatform/kubernetes/pull/10957/files#diff-77>
   (210)

Patch Links:

   - https://github.com/GoogleCloudPlatform/kubernetes/pull/10957.patch
   - https://github.com/GoogleCloudPlatform/kubernetes/pull/10957.diff

—
Reply to this email directly or view it on GitHub
<https://github.com/GoogleCloudPlatform/kubernetes/pull/10957>.

liggitt commented on plugin/pkg/auth/authenticator/token/openid/openid.go:
```diff
@@ -0,0 +1,112 @@
+		return nil, false, err
+	}
+
+	name, ok, err := claims.StringClaim("name")
```
http://openid.net/specs/openid-connect-core-1_0.html#StandardClaims
the standard subject claim ("sub") should be used... "name" is the display name, and is not likely to be unique or immutable.

yifan-gu replied:
Thanks for catching it :)

liggitt commented on plugin/pkg/auth/authenticator/token/openid/openid.go:
```diff
@@ -0,0 +1,112 @@
+	if err != nil {
+		return nil, err
+	}
+	client.SyncProviderConfig(discovery)
```
does this mean the sync only happens when the server starts? some OIDC providers rotate keys daily

liggitt replied:
nevermind, I see it syncs regularly. what is the default frequency?

yifan-gu replied:
@liggitt It depends on the [exp of the response](https://github.com/coreos/go-oidc/blob/master/oidc/provider.go#L117), and has a [Max and Min value](https://github.com/coreos/go-oidc/blob/master/oidc/provider.go#L22)

yifan-gu replied:
Will add some comments for this.

liggitt commented:
I like how self contained this is, but it's a little hard to understand how you would actually obtain a token to use against the API this way

liggitt commented on plugin/pkg/auth/authenticator/token/openid/openid.go:
```diff
@@ -0,0 +1,116 @@
+	}
+
+	ccfg := oidc.ClientConfig{
+		Credentials:    oidc.ClientCredentials{ID: clientID},
```
what is the clientID used for? we're not requesting tokens...

yifan-gu replied:
@liggitt clientID is for claims verification, we need to [verify the 'aud'](http://openid.net/specs/openid-connect-core-1_0.html#IDTokenValidation)

liggitt commented on plugin/pkg/auth/authenticator/token/openid/openid.go:
```diff
@@ -0,0 +1,116 @@
+
+	ccfg := oidc.ClientConfig{
+		Credentials:    oidc.ClientCredentials{ID: clientID},
+		ProviderConfig: cfg,
```
many installations will need a way to specify a CA bundle when pointing to an internal openid server

yifan-gu replied:
@liggitt Ahh, will add it, thanks!

liggitt commented on pkg/apiserver/authn.go:
```diff
@@ -28,11 +28,12 @@ import (
 
 // NewAuthenticator returns an authenticator.Request or an error
-func NewAuthenticator(basicAuthFile, clientCAFile, tokenFile, serviceAccountKeyFile string, serviceAccountLookup bool, helper tools.EtcdHelper) (authenticator.Request, error) {
+func NewAuthenticator(basicAuthFile, clientCAFile, tokenFile, openIDClientID, openIDEndpoint, serviceAccountKeyFile string, serviceAccountLookup bool, helper tools.EtcdHelper) (authenticator.Request, error) {
```
I probably should have done it when I added the service account options, but this is probably at the point where we need an Options struct with sub-structs for each authenticator type.

Basic auth: basicAuthFile
Cert auth: clientCAFile
CSV token: tokenFile
OpenID token: openIDClientID, openIdEndpoint
Service account token: serviceAccountKeyFile, serviceAccountLookup, helper

yifan-gu replied:
Totally agree.

philips commented:
@liggitt You would obtain a token out of band and place it into the kubectl configuration. It would be great to add a nice interface for `kubectl` to do the browser auth workflow and write-out/update the kubectl configuration file like the gcloud SDK does, but one step at a time.

yifan-gu commented:
@liggitt Added cert bundles to enable SSL connection to the openid provider. Refactored `NewAuthenticator` to take a struct instead of bunch of flags. And added related tests.

liggitt commented on cmd/kube-apiserver/app/server.go:
```diff
@@ -178,6 +181,9 @@ func (s *APIServer) AddFlags(fs *pflag.FlagSet) {
 	fs.StringVar(&s.TokenAuthFile, "token-auth-file", s.TokenAuthFile, "If set, the file that will be used to secure the secure port of the API server via token authentication.")
+	fs.StringVar(&s.OpenIDEndpoint, "openid-endpoint", s.OpenIDEndpoint, "If set, the endpoint will be used to verify the JWT using OpenID protocol")
+	fs.StringVar(&s.OpenIDClientID, "openid-client-id", s.OpenIDClientID, "The client ID for the OpenID client, must be set if openid-endpoint is set")
+	fs.StringVar(&s.OpenIDServerCertFile, "openid-server-cert", s.OpenIDServerCertFile, "The PERM-encoded x509 certificate for the OpenID server, optional")
```
`PEM-encoded`, `OpenIDCAFile`, `openid-ca-file` (to follow client-ca-file pattern above)

liggitt commented on pkg/apiserver/authn.go:
```diff
@@ -103,6 +123,16 @@ func newAuthenticatorFromTokenFile(tokenAuthFile string) (authenticator.Request,
 
+// newAuthenticatorFromOpenIDEndpoint returns an authenticator.Request or an error.
+func newAuthenticatorFromOpenIDEndpoint(clientID, endpoint, certFile string) (authenticator.Request, error) {
+	tokenAuthenticator, err := openid.NewOpenID(clientID, endpoint, certFile)
```
plumb `caFile` throughout

liggitt commented on plugin/pkg/auth/authenticator/token/openid/openid.go:
```diff
@@ -0,0 +1,147 @@
+	var roots *x509.CertPool
+
+	if certFile == "" {
+		glog.Warning("No x509 certificate provided, will establish insecure connection between the OpenID service")
```
If there are no roots, it should use system roots to verify the connection, which is not insecure...

liggitt commented:
@yifan-gu updates look good, thanks. Just a few comments on the naming, param name, and warning

yifan-gu commented:
@liggitt Comments addressed, thanks! And reworked the warning message/logic to make it clearer.

liggitt commented:
Authn changes LGTM, though I didn't look through the new godeps or try it out yet. Is there a relevant doc that could get a sentence or two with pointers to obtaining tokens for this authenticator?

yifan-gu commented:
@liggitt Not sure if these links help, but here is the [godoc](https://godoc.org/github.com/coreos/go-oidc/oidc) of the library, and here is an [example app](https://github.com/coreos/go-oidc/blob/master/example/app/main.go) using library.

Basically the token is an [id token](http://openid.net/specs/openid-connect-core-1_0.html#IDToken) signed by the authorization server.

liggitt commented:
> Basically the token is an id token signed by the authorization server. It is a JWT.

Sure, but as a user, it's really hard to get an ID token from an OIDC provider which is valid for the `client_id` the Kubernetes server is configured with. If the expectation is that the cluster administrator has a separate app that walks users through obtaining the token, that should probably be mentioned, otherwise as a user, I might think I was supposed to be able to get the token on my own.

A paragraph around the OpenID token auth should go in docs/admin/authentication.md

yifan-gu commented:
@liggitt ACK.

yifan-gu commented:
@liggitt Added some docs and explanation. PTAL, thanks!

liggitt commented on docs/admin/authentication.md:
```diff
@@ -51,6 +50,22 @@ and is a csv file with 3 columns: token, user name, user uid.
+- `--openid-client-id` (required) is used by apiserver to verify the audience of the token.
+A valid [ID token](http://openid.net/specs/openid-connect-core-1_0.html#IDToken) MUST have this
+client-id in its `aud` claims.
+- `--openid-ca-file` (optional) is used by apiserver to establish and verify the security connection
```
s/security/secure/

liggitt commented on docs/admin/authentication.md:
```diff
@@ -51,6 +50,22 @@ and is a csv file with 3 columns: token, user name, user uid.
+to the OpenID provider.
+
+Currently, the ID token will be obtained by some third-party app. This means the app and apiserver
+need share the `--openid-client-id`. We expect `kubectl` be able to get the token soon, as well as
```
s/need/must/
"...to be able to"

liggitt commented on docs/admin/authentication.md:
```diff
@@ -51,6 +50,22 @@ and is a csv file with 3 columns: token, user name, user uid.
+
+Currently, the ID token will be obtained by some third-party app. This means the app and apiserver
+need share the `--openid-client-id`. We expect `kubectl` be able to get the token soon, as well as
+to refresh the token with [refresh token](http://openid.net/specs/openid-connect-core-1_0.html#RefreshTokens)
```
"...with a refresh token…"

liggitt commented:
Just a couple nits, the doc paragraph is helpful, thanks.

yifan-gu commented:
@liggitt Thanks :)

uruddarraju commented:
Any idea when this can be merged ?

yifan-gu commented:
@uruddarraju We would need a lgtm from a committer (/cc @erictune @smarterclayton ) , then I think we should be able to merge this when the code freeze is over.

yifan-gu commented:
@erictune Any feedback on this one :) ?

erictune commented:
Mostly happy with this.  Just want to make sure that I understand how:
 - we will be able to show a human-readable identifier for users in a future UI.
 - we will be able to write policy using human-readable user strings, or else convince me that is the wrong model
 - we compose use user-strings in a way that is consistent.

yifan-gu commented:
@liggitt @ericelliott Sorry, the new rebase commits purge your comments on old commits...

erictune commented:
@yifan-gu you meant @erictune

yifan-gu commented:
> we will be able to show a human-readable identifier for users in a future UI.
we will be able to write policy using human-readable user strings, or else convince me that is the wrong model
we compose use user-strings in a way that is consistent.

In this case I lean towards using `name` as user name,  and `sub` as user id. Can we add something like `GetUserID()` in the [authorizer.Attributes interface](https://github.com/GoogleCloudPlatform/kubernetes/blob/master/pkg/auth/authorizer/interfaces.go#L25) later so this can solve the unique identity problem, (e.g. call `GetUserID()` first, if user id is empty, falls through `GetUserName()`).

liggitt commented:
the name returned by UserInfo.GetName must be unique at that point in time. We can't use possibly non-unique values (like a display name) for it.

yifan-gu commented:
@liggitt OK, but assuming we have UserID, why do we still need unique user name?

yifan-gu commented:
It will be perfect if we can have unique user name, which is also readable. But if we don't, we can use a unique id to identify the user. I am thinking we can still show and use the non-unique user name to write policies as an option, in case if there are conflicts, we would require a user id... (maybe that seems complicated)

liggitt commented:
Name and UID in the UserInfo interface hold the same meaning as Name/UID in API objects. name uniquely identifies the user at that point in time. UID uniquely identifies the user forever, to distinguish between deleted and recreated users of the same name. 

I think the name used to write policy against needs to be more deterministic… I would be very uncertain granting permissions to something that could match either a human readable username or a unique numeric id. What if someone's human readable name conflicts with a unique numeric id?

yifan-gu commented:
> I think the name used to write policy against needs to be more deterministic… I would be very uncertain granting permissions to something that could match either a human readable username or a unique numeric id. What if someone's human readable name conflicts with a unique numeric id?

I understand your concern. For writing policy particularly, I was trying to say that by default we do something like `SELECT name FROM *`. If there is no conflicts, we are happy. Otherwise we will receive an error, and need to do `SELECT name, id FROM *`. We won't have human readable name conflicts with the id as long as we don't do `SELECT id FROM *` in the above case.

In other words, we can view `name` as REQUIRED, and `id` as OPTIONAL.

> Name and UID in the UserInfo interface hold the same meaning as Name/UID in API objects. name uniquely identifies the user at that point in time. UID uniquely identifies the user forever, to distinguish between deleted and recreated users of the same name.

Thanks for heading up on this:) I am not clear how this will play out in our case, need to think about it.

jboeuf commented:
@erictune asked me to have a look at this PR.

jboeuf commented on plugin/pkg/auth/authenticator/token/openid/openid.go:
```diff
@@ -0,0 +1,156 @@
+	client *oidc.Client
+}
+
+// NewOpenID creates a new OpenID client with the given clientID and discovery URL.
```
I think that renaming 'discovery' to 'issuerURL' would be closer to the spec (and the underlying library):
https://openid.net/specs/openid-connect-discovery-1_0.html#ProviderConfig
IMO it is also easier to understand for the client of the library.

yifan-gu replied:
ack.

jboeuf replied:
Thanks. So do you plan to change it or not? After the exchange below regarding http vs https, I think that it is actually important.

jboeuf commented on plugin/pkg/auth/authenticator/token/openid/openid.go:
```diff
@@ -0,0 +1,156 @@
+	}
+
+	if url.Scheme == "http" {
+		glog.Warning("Using insecure connection to the OpenID service")
```
Are we sure that we want to allow this? If we don't have a good use case for it, we should just not support it as it is clearly a security issue.

yifan-gu replied:
@jboeuf When I did this, I was thinking about some behind-firewall scenario, e.g. where the apiserver and the provider stay in the same behind-firewall network. I am open to be convinced to use ssl in any cases though.

jboeuf replied:
@yifan-gu thanks for the reponse.

Here are a few arguments I hope to convince you with.

1. Allowing http is actually a spec violation. See section 3. of https://openid.net/specs/openid-connect-discovery-1_0.html
> issuer
> REQUIRED. URL using the ***https*** scheme with no query or fragment component that the OP asserts as its Issuer Identifier. If Issuer discovery is supported (see Section 2), this value MUST be identical to the issuer value returned by WebFinger. This also MUST be identical to the iss Claim value in ID Tokens issued from this Issuer.

2. I don't think that the protection of a firewalled network is something that you want to rely on. It means that anybody that has access to your network can MITM this http exchange. With HTTPS, an attacker would need to have access to the keys on the machine which significantly reduces the surface of attack.

3. If customers really want this "feature", I think that you should at the very least protect it with a flag or some kind of global config so that the said customers avoid shooting themselves in the foot.

What do you think?

liggitt replied:
+1000 for requiring https, per spec and per us security types sleeping at night

jboeuf commented on plugin/pkg/auth/authenticator/token/openid/openid.go:
```diff
@@ -0,0 +1,156 @@
+		return nil, false, fmt.Errorf("cannot find user subject in JWT claims")
+	}
+
+	// TODO(yifan): Add UID and Group.
```
I think that adding the issuer would be a good thing as well as the upper layer that consume these objects (e.g. authorization) may (should actually) want to know who is asserting this identity.

yifan-gu replied:
@jboeuf Yea, I can add a TODO here, I think we can do it in following PRs if we want, as we can also get issuer info from cert based authenticators.

yifan-gu commented:
@jboeuf Thanks for the review! I made changes according to your comments, please take a look :)

yifan-gu commented:
The tests are failing because some is not using ssl. Will fix tomorrow.

jboeuf commented:
@yifan-gu thanks much for the fixes and the TODO for exposing the issuer to the upper layers. LGTM.

From the comments above, you may want to think about how you're going to write policies based on the subject and the issuer in another PR. Maybe you could construct a canonical identifier based on these 2 parameters.

yifan-gu commented:
@liggitt @erictune @jboeuf 
Want to give another thought here:
Though the OIDC spec says we should not rely on email to be unique. But in practice I think changing email is not common, so maybe it's ok to just update it when his/her email changes? To me this seems like an feasible solution for now, until we have something better. Any ideas?

liggitt commented:
There are two issues, uniqueness and immutability. Email satisfies neither, which would allow a user to lose their permissions if their email changed, or allow a user to spoof another user's permissions, simply by having the same email address (even if unverified). I see the following options:

1. Use a unique id (issuer+subject claim)
2. Same as 1, but allow the claim to be configured (putting the decision of trusting email addresses from the provider on the cluster admin)
3. Build an identity from the issuer+subject, and map it to a user whose username is the email or preferred_login (other configurable claim). This requires persisted objects in etcd to remember the identity and user, but protects against spoofing (only the first one in gets the preferred username) and mutability (the mapping persists even if the email changes). This is a bigger change, out of scope for this PR. Fwiw, that's where OpenShift ended up with our auth integration stuff.

yifan-gu commented:
/cc @bcwaldon @bobbyrullo

jboeuf commented:
@liggitt +1000 on this one :).

yifan-gu commented:
@liggitt Thanks for the reply :) but I am not sure I understand the 2nd option here.

Apart from that,
For spoofing, can we solve it by requiring that email **MUST** be verified? 
For immutability, I would still argue that we can assume this is not a frequent event. Besides, in some cases when a user changes his email address (e.g. his company email address), this might imply the permission _really_ needs to be updated as well?

Thanks :)

liggitt commented:
For the scope of this PR, I don't think it's a good idea to default to anything other than the `sub` claim (possibly with the issuer as a prefix) as the default username.

I think it's reasonable to let the cluster admin choose which claim to use as the username, and put the burden on them to live with the consequences of ignoring the OpenID spec statements about email and preferred_username's uniqueness and immutability.

yifan-gu commented:
> I think it's reasonable to let the cluster admin choose which claim to use as the username, and put the burden on them to live with the consequences of ignoring the OpenID spec statements about email and preferred_username's uniqueness and immutability.

This is fair, but please pardon me I am not sure how this looks like? How could admin choose which to use?

liggitt commented:
`--oidc-username-claim`, defaults to `"sub"`?

yifan-gu commented:
Gotha, thanks. I will update the PR :)

liggitt commented:
Before you do that, I'd like @jboeuf's opinion on that approach... just want to save you time reworking if there's disagreement

erictune commented:
@liggitt in openshift, what is [RoleBinding.username](https://github.com/openshift/origin/blob/89c256af5407fecdf142f2d1db752199bfd2d18a/pkg/authorization/api/v1/types.go) when using OpenID for authentication?  Is it an email, or is it a `sub` claim, or something else?  If it is a `sub` claim, how are humans supposed to read a `RoleBinding`? Or do they use some other UI to generate/edit `RoleBinding`s which shows the username that the uids map to?

liggitt commented:
In OpenShift, we have Identity and User objects. An Identity maps to a single User, and multiple Identities can map to the same User. The RoleBinding.Username is the User.Name. The separation of Identity and User lets us map ugly unambiguous values like an OpenID subject claim to nice human-readable values (like a preferred_username claim) that you can write policy against.

When you log in using an identity provider (OpenID, LDAP, auth proxy, etc), we get three pieces of information:
1. A unique provider name (part of cluster auth config, e.g. "google-openid", "acme-ldap")
2. A unique identity within the scope of that provider (e.g. the "sub" claim for OpenID, the "uid" attribute for an LDAP provider).
3. A preferred username (defaults to the unique identity, but can be something different like "preferred_username" or "email" for OpenID, or a "login" attribute for LDAP)

We use 1 and 2 to make the Identity.Name (e.g. "google-openid:123456", "acme-ldap:liggitt").

If that Identity already exists, we find the mapped User and we're done.

If that Identity does not yet exist, we create it, use the preferred username to create a User, and map the new Identity to the new User. If a User with that preferred username already exists, we have several options:
1. Fail
2. Deconflict the preferred username (the second "bob" to log in gets "bob2")
3. Map the new Identity to the existing User in addition to what was already mapped

More details at https://docs.openshift.org/latest/architecture/additional_concepts/other_api_objects.html#user-objects

erictune commented:
@liggitt 

Is there a concern that an admin who is responsible for setting policy might see a username, such as `liggitt` and assume it is a certain person, e.g. `redhat-ldap:liggitt` and put it in a RoleBinding, when in fact it is the wrong person, e.g. `evilcorp-openid:liggitt`, who happened to get there first?

erictune commented:
@liggitt 
broader questions:
- In your opinion, should we upstream User and Identity?  
- Would someone from RedHat have time to do that and in what time frame?
- Does accepting this PR now make the above much harder?

@yifan-gu 
if we took the Red Hat OpenID solution instead of this one, would it work for your use cases?
If not why not?

erictune commented:
happy to have longer convo offline if needed.  You can email me.  I am username etune.  I am at domain google dot com.

yifan-gu commented:
> Is there a concern that an admin who is responsible for setting policy might see a username, such as liggitt and assume it is a certain person, e.g. redhat-ldap:liggitt and put it in a RoleBinding, when in fact it is the wrong person, e.g. evilcorp-openid:liggitt, who happened to get there first?

From my understanding on https://github.com/GoogleCloudPlatform/kubernetes/pull/10957#issuecomment-128413461 If `redhat-ldap:liggitt` and `evilcorp-openid:liggitt` are mapped to one `liggitt`, it means they are the same person, otherwise we will get `liggitt` and `liggitt2`, is that correct? @liggitt

erictune commented:
Seems like that depended on whether the admin chose the Fail, deconflict, or map-to-existing policy?

yifan-gu commented:
> if we took the Red Hat OpenID solution instead of this one, would it work for your use cases?
If not why not?

AFAICS, I am fine as long as the api-server could authenticate an OIDC token against a provider.
FWIW, we will have follow-up patches to `kubectl` that to handle token expiration and refresh, but that shouldn't matter.
/cc @bobbyrullo @bcwaldon

liggitt commented:
> Is there a concern that an admin who is responsible for setting policy might see a username, such as liggitt and assume it is a certain person, e.g. redhat-ldap:liggitt and put it in a RoleBinding, when in fact it is the wrong person, e.g. evilcorp-openid:liggitt, who happened to get there first?

We strongly recommend not configuring EvilCorp as a trusted identity provider ;-)
It hasn't been an issue yet in practice, but if it became one, there are two other policies that could be created:
* Lookup-only (admin has to create identity -> user mapping, prevents a username "land grab")
* Fully-qualified usernames (username includes the provider name as a prefix, or suffix, etc)

You could even have a different strategy per identity provider, so "acme-ldap" users get their usernames automatically, and evilcorp users have to be reviewed and mapped to a user by an admin before they can log in.

> In your opinion, should we upstream User and Identity?
> Would someone from RedHat have time to do that and in what time frame?

I like the flexibility of the Identity->User model. I think it's worth upstreaming. It would probably either be @deads2k or I. For the moment, I'd rather focus on the authorization proposal and interfaces, then move to this. Also, since we do our identity->user transformation when users log in and get their OAuth tokens, so upstreaming would probably involve discussing Kubernetes as an OAuth provider (which is fine, but might not be short).

> Does accepting this PR now make the above much harder?

I don't think this PR makes it harder... the interactions with k8s code are minimal. I am wary of `kubectl` adding capabilities around obtaining/renewing tokens from non-kubernetes sources... that seems a likely follow-up to this PR, and I'm not sure that makes sense.

liggitt commented on docs/admin/authentication.md:
```diff
@@ -52,6 +51,22 @@ and is a csv file with 3 columns: token, user name, user uid.
+to the OpenID provider.
+
+Currently, the ID token will be obtained by some third-party app. This means the app and apiserver
+MUST share the `--oidc-client-id`. We expect `kubectl` to be able to get the token soon, as well as
```
> We expect `kubectl` to be able to get the token soon

Really? How would you drive an OAuth flow against an arbitrary OIDC provider from the command line, to get a token which was valid for the `client_id` the apiserver is configured with, without having access to the `client_secret`?

yifan-gu replied:
@liggitt Seems i'd better delete this sentence.

liggitt commented on docs/admin/authentication.md:
```diff
@@ -52,6 +51,22 @@ and is a csv file with 3 columns: token, user name, user uid.
+Currently, the ID token will be obtained by some third-party app. This means the app and apiserver
+MUST share the `--oidc-client-id`. We expect `kubectl` to be able to get the token soon, as well as
+to refresh the token with a [refresh token](http://openid.net/specs/openid-connect-core-1_0.html#RefreshTokens)
+automatically when the token expires.
```
Making kubectl talk to arbitrary identity servers feels like it might be straying a bit. I think I'd rather eventually see OAuth capabilities added to kubernetes in a way that kubectl could work against a known implementation, and the kubernetes server implementation could facilitate obtaining the OIDC token from the third-party.

yifan-gu commented:
> Really? How would you drive an OAuth flow against an arbitrary OIDC provider from the command line, to get a token which was valid for the client_id the apiserver is configured with, without having access to the client_secret?

Previously, I imagine the workflow looks like `gcloud` tools, which has the a `client_id` and `client_secret` hardcoded in the config. And the apiserver will share this `client_id`. But one of the problems is in some cases it's not feasible to open a web browser to do the login. So before we decide we really want to do that, for now we can just get the `id_token`, `refresh_token` from some place, as well as the `client_id`, `client_secret`, and copy-paste them in `kubeconfig`.

For the refreshing flow, the kubectl will try to refresh the id_token if it expires, or authentication fails.

yifan-gu commented:
@erictune @jboeuf @liggitt Any objections on a solution based on https://github.com/GoogleCloudPlatform/kubernetes/pull/10957#issuecomment-127707717 and https://github.com/GoogleCloudPlatform/kubernetes/pull/10957#issuecomment-128095808 ?

erictune commented:
@yifan-gu 
After this was merged, what were you planning to do for authorization?

bobbyrullo commented:
@erictune - our initial plan is to just use hand written Kubernetes JSON policy files, where the user ID maps back to the OpenID Connect "sub" claim. Beyond that, the path a little less clear, and is open for discussion. Something based on some standard for making authz requests would be ideal, but I'm not sure what that would be; one idea would be settling on return user policies in JWT claims, but we haven't really thought that out fully yet.

jboeuf commented:
@bobbyrullo Is the mapping from user ID to just `sub` or to `sub` + `iss`? I like the latter and I'm concerned with the former.

bobbyrullo commented:
Sure, that leaves open the possibility of authenticating against multiple providers.

jboeuf commented:
Perfect then. As long as the authenticated principal referenced (possibly indirectly) in the ACLs is sub+iss, I'm happy :).

yifan-gu commented:
@jboeuf Cool, I will update the PR to take `issuer:sub` as default user ID. Admin can also choose `issuer:email` or `issuer:username` which is not guaranteed to be unique and immutable according to the oidc spec, but more readable and in practice should be _pretty_ unique and immutable (e.g. email).

According to the [definition](http://openid.net/specs/openid-connect-discovery-1_0.html#IssuerDiscovery), issuer is a URI() which can contains `:ports`, so this can result in something like `example.com:8000:id-foo` but I don't think this is a problem? Or do you have something better to suggest? Thanks :)

liggitt commented:
the option to have fully-qualified issuer+sub usernames is good to have, but in practice, most installations will not need such specificity, and having to express ACL rules like "https://accounts.google.com:1234567890" is almost impossible to use (especially once end-users are able to add ACL grants). As long as a cluster admin can choose between something like "issuer:sub" format and "email", I think our bases are covered.

jboeuf commented:
@liggitt Isn't that a UI issue? When editing ACLs, end users could be searching for friendly names that map to fully qualified principals (iss + sub) and iss+sub is what would be in the actual ACL.

@yifan-gu Since `iss` cannot have fragments, how about `#` as a separator: that would disambiguate  between `example.com:8000#id-foo` and `example.com#8000:id-foo`.

liggitt commented:
if we ever persist identities or users, the included `/` and `#` characters will cause us grief if they are in the identity or user name

jboeuf commented:
Sorry @liggitt, can you give me an example? I'm not sure I understand. Note that the `iss` can have a path component.

liggitt commented:
> Sorry @liggitt, can you give me an example?

etcd would be unhappy storing an object whose key was "https://accounts.google.com#12345"

jboeuf commented:
OK, i had a quick look look at etcd and it uses the key as part of the URL in the API. But if it does so, it would seem dangerous to not URL-encode these keys before using them as part of a REST API in etcd no matter what these are, right? Also, I don't think that the leading `https://` is necessary since it is a given for issuer (but that does not really matter).

liggitt commented:
> OK, i had a quick look look at etcd and it uses the key as part of the URL in the API. But if it does so, it would seem dangerous to not URL-encode these keys before using them as part of a REST API in etcd no matter what these are, right?

Yeah, it causes problems both in the REST API URL and in the etcd API URL. Also, URL-encoding path segments doesn't work well in Go (Go helpfully decodes them for you so you can't easily tell the difference between `/foo%2Fbar` and `/foo/bar`).

jboeuf commented:
@liggitt This seems broken to me :(. Maybe we could chat offline about this. I am username jboeuf in the email domain google dot com. 

Regardless of the separator, the `iss` field may contain a `/` and I don't know what to do about this one.

yifan-gu commented:
Do we care about what will end up in the persistent store? If not, can we hash it or even base64 encode it?

liggitt commented:
it's not just the persistent store, it would be the name of the objects in the REST API as well. both make the name part of a URL, so we have to be careful about what we make part of names of things that might end up being API objects

jboeuf commented:
So... where does that leave us? The `sub` claim is also free form so may contain problematic characters. Should the unique identifiers for principals be `UrlBase64(iss | '#' | sub)`? where `|` denotes concatenation?

yifan-gu commented:
> it's not just the persistent store, it would be the name of the objects in the REST API as well. both make the name part of a URL, so we have to be careful about what we make part of names of things that might end up being API objects

Actually does the user name has any constraints right now? AFAICS it's just a plain string, so I think we will need to encode them anyway? @liggitt @jboeuf

Commit:
plugin/pkg/auth: add OpenID Connect token authenticator.

Also add related new flags to apiserver:
"--oidc-issuer-url", "--oidc-client-id", "--oidc-ca-file", "--oidc-username-claim",
to enable OpenID Connect authentication.

Commit:
plugin/pkg/auth: add tests for OpenID Connect authenticator.

Commit:
docs: add documentations for OpenID Connect authentication.

yifan-gu commented:
@liggitt @jboeuf @erictune 
Get back on this one. I added `--oidc--username-claim` to let the admin choose which claim to use as the user name. Also added documentation around this. in [authencitation.md](https://github.com/yifan-gu/kubernetes/blob/b98dbc13431d571d9743d6ea669463571dfa2e35/docs/admin/authentication.md#authentication-plugins)
By default, the `issuerURL` + `sub` will be used, (e.g. `example.com#user-foo`), if `--oidc-username-claim="email"`, then username will be the email address (e.g. `user-foo@example.com`)

Any objection or feedback? Thank you :)

yifan-gu commented:
Might need to update the godep commit as it includes some nonrelated changes.

jboeuf commented:
@yifan-gu: sounds good to me.

liggitt commented:
sounds good to me as well. I wonder if a note in the docs similar to https://github.com/kubernetes/kubernetes/pull/11798#issuecomment-129655212 would be useful, until we settle more on how we want to handle any identity/user pieces.

yifan-gu commented:
@liggitt Good suggestion, docs added :)
@erictune Any more advice? Seems we are close to the finish line :stuck_out_tongue: :stuck_out_tongue:

yifan-gu commented:
Will need to rebase after https://github.com/kubernetes/kubernetes/pull/12966

yifan-gu commented:
Godeps update. ping @erictune :)

erictune commented:
@jboeuf @yifan-gu 
Near the top of https://developers.google.com/identity/protocols/OpenIDConnect?hl=en#getcredentials it says that you need to have a client secret to use OpenID connect.
IIUC, in this case, "client" means the kubernetes apiserver.
I don't see a flag that lets the apiserver know what it's client secret is.

Am I misunderstanding, or is the use of the client secret something that just Google does, and this PR is still useful for flows that don't require a client secret, or what?

erictune commented:
Maybe the client secret is only needed if requesting scopes other than oauth?  Docs are confusing.

liggitt commented:
client_secret is only required to obtain the token from the OIDC provider. This authenticator isn't obtaining the token, it is just validating it

erictune commented:
LGTM.

mikedanese commented:
@k8s-bot ok to test.

Commit:
Godeps: add Godpes for go-oidc.

zmerlynn commented:
@k8s-bot test this

erictune commented:
Test failures are:
```
[Fail] DNS [It] should provide DNS for the cluster 
/go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/test/e2e/dns.go:198

[Fail] DNS [It] should provide DNS for services 
/go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/test/e2e/dns.go:236
```
